### ISSUE TITLE
Import error in my Angular project

### ISSUE THREAD

**Issue by weilies** [guigrpa/docx-templates#207]
I have the import script from the sample
```
...
import createReport from 'docx-templates';
import fs from 'fs';
...
```

but getting error in browser
```
Failed to compile.

node_modules/docx-templates/lib/main.d.ts:3:8 - error TS1259: Module '"/home/weilies_chok/apps/my-app/node_modules/jszip/index"' can only be default-imported using the 'allowSyntheticDefaultImports' flag

3 import JSZip from 'jszip';
         ~~~~~

  node_modules/jszip/index.d.ts:270:1
    270 export = JSZip;
        ~~~~~~~~~~~~~~~
    This module is declared with using 'export =', and can only be used with a default import when using the 'allowSyntheticDefaultImports' flag.
```

compiler error
```
✔ Browser application bundle generation complete.

5 unchanged chunks

Build at: 2021-04-29T03:04:50.495Z - Hash: 54914e41f6300bfbc2a7 - Time: 476ms

Error: node_modules/docx-templates/lib/main.d.ts:3:8 - error TS1259: Module '"/home/weilies_chok/apps/my-app/node_modules/jszip/index"' can only be default-imported using the 'allowSyntheticDefaultImports' flag

3 import JSZip from 'jszip';
         ~~~~~

  node_modules/jszip/index.d.ts:270:1
    270 export = JSZip;
        ~~~~~~~~~~~~~~~
    This module is declared with using 'export =', and can only be used with a default import when using the 'allowSyntheticDefaultImports' flag.


Error: src/app/app.component.ts:3:8 - error TS1192: Module '"fs"' has no default export.

3 import fs from 'fs';
         ~~
```

package.json
```
{
  "name": "my-app",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "~11.2.4",
    "@angular/common": "~11.2.4",
    "@angular/compiler": "~11.2.4",
    "@angular/core": "~11.2.4",
    "@angular/elements": "^11.2.4",
    "@angular/forms": "~11.2.4",
    "@angular/platform-browser": "~11.2.4",
    "@angular/platform-browser-dynamic": "~11.2.4",
    "@angular/router": "~11.2.4",
    "angular-formio": "^4.11.4",
    "bootstrap": "^4.6.0",
    "docx-templates": "^4.6.0",
    "fs": "*",
    "nativescript-ibeacon": "^0.8.1",
    "rxjs": "~6.6.0",
    "tslib": "^2.0.0",
    "zone.js": "~0.11.3"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.1102.3",
    "@angular/cli": "~11.2.3",
    "@angular/compiler-cli": "~11.2.4",
    "@types/jasmine": "~3.6.0",
    "@types/node": "^12.11.1",
    "codelyzer": "^6.0.0",
    "jasmine-core": "~3.6.0",
    "jasmine-spec-reporter": "~5.0.0",
    "karma": "~6.1.0",
    "karma-chrome-launcher": "~3.1.0",
    "karma-coverage": "~2.0.3",
    "karma-jasmine": "~4.0.0",
    "karma-jasmine-html-reporter": "^1.5.0",
    "protractor": "~7.0.0",
    "ts-node": "~8.3.0",
    "tslint": "~6.1.0",
    "typescript": "~4.1.5"
  }
}

```

**Comment by jjhbw** [comment]
This is because the Angular compiler doesn't polyfill the required packages, contrary to webpack. Related to #188  #135 .

**Comment by weilies** [comment]
hi @jjhbw thanks for the respond and may i know if the team is preparing to fix it? Or  i hit a deadend? :) 
i basically can't find a replacement for such a great package so far lol

**Comment by jjhbw** [comment]
Hi @weilies , no i'm afraid I don't know how to fix it. This seems like an issue specific to the Angular toolchain, and I expect it to be pretty common. I think you're better off looking for Angular-specific fixes. I have never worked with the Angular toolchain before, so I wouldn't know what to do.

If I can do something to improve the ergonomics of docx-templates for Angular users, just let me know in this thread and I'll try to help.

**Comment by mathe42** [comment]
Just some ideas:

1. add allowSyntheticDefaultImports: true to your tsconfig
2. DON'T USE `fs` this is not useable in browsers (so unless your writing a Angular page for electron don't use fs)

Instead of fs you have to use the `fetch` api to get the .docx file from your server or use `<input type="file">` to get the file or the new Native Filesystem API (see https://caniuse.com/?search=native%20file )

**Comment by weilies** [comment]
> Just some ideas:
> 
> 1. add allowSyntheticDefaultImports: true to your tsconfig
> 2. DON'T USE `fs` this is not useable in browsers (so unless your writing a Angular page for electron don't use fs)
> 
> Instead of fs you have to use the `fetch` api to get the .docx file from your server or use `<input type="file">` to get the file or the new Native Filesystem API (see https://caniuse.com/?search=native%20file )

Thanks. It's inspiring but i have to wait for some guru to make a POC LOL. I am not a hard core developer but more a solution integrator. Without a working sample, what you mentioned r a bit too much for me. But thanks!!

**Comment by mathe42** [comment]
Here a small example how this might looks (useing fetch) so it might help...
https://github.com/EC-Nordbund/EC-Verwaltung/blob/old-version/src/report.ts

**Comment by weilies** [comment]
```
  async  postData(url = '', data = {}) {
    const template = await fetch(url).then((v) => v.arrayBuffer());
    const report:ArrayBuffer = await createReport({
      data,
      template    ==============> ERROR
    });
    this.saveByteArray('myfile.docx', report, true);


  }

  onPostData() {
    let fURL = 'https://firebasestorage.googleapis.com/v0/b/fx-vps-225216.appspot.com/o/tag-example-mod.docx?alt=media&token=3da1cc23-339c-447a-a9e0-e57cc6788155'
    this.postData(fURL, { answer: 42 })
  }

  saveByteArray(reportName: string, byte: ArrayBuffer, word: boolean = false) {
    const blob = new Blob([byte], {
      type: (
        word ? 'vnd.openxmlformats-officedocument.wordprocessingml.document'
          : 'vnd.openxmlformats-officedocument.spreadsheetml.sheet'
      )
    });
    const link = document.getElementById('doc-download') as HTMLAnchorElement;
    link.href = window.URL.createObjectURL(blob);
    link.download = reportName;
    link.click();
  }
```

I am getting error 

> Docx file template as a NodeJS Buffer or Buffer-like object in Browsers.
> 
> Type 'ArrayBuffer' is missing the following properties from type 'Buffer': readBigUInt64BE, readBigUInt64LE, readBigInt64BE, readBigInt64LE, and 75 more.ts(2740)
> types.d.ts(39, 5): The expected type comes from property 'template' which is declared here on type 'UserOptions'


I am able to fetch the content of the doc

> ...
> [[Int8Array]]: Int8Array(40611)
> [[Uint8Array]]: Uint8Array(40611)
> [[ArrayBufferByteLength]]: 40611
> [[ArrayBufferData]]: "0x1b2c05bbc000"

**Comment by mathe42** [comment]
const template: Buffer = await fetch(...)

**Comment by weilies** [comment]
const template: Buffer = await fetch(url).then((v) => v.arrayBuffer());

error

> Error: src/app/app.component.ts:105:11 - error TS2322: Type 'Buffer | ArrayBuffer' is not assignable to type 'Buffer'.
>   Type 'ArrayBuffer' is missing the following properties from type 'Buffer': readBigUInt64BE, readBigUInt64LE, readBigInt64BE, readBigInt64LE, and 75 more.
> 
> 105     const template: Buffer = await fetch(url).then((v) => v.arrayBuffer());

**Comment by mathe42** [comment]
const template: Buffer = await fetch(url).then((v) => v.arrayBuffer() as Promise<any>);

**Comment by weilies** [comment]
The code is now without syntax error but returned with build error lol

> Build at: 2021-05-08T10:10:58.912Z - Hash: 57c768fb928842240761 - Time: 2816ms
> 
> Warning: ./node_modules/sax/lib/sax.js
> Module not found: Error: Can't resolve 'stream' in '/home/weilies_chok/apps/my-app/node_modules/sax/lib'
> 
> 
> 
> Error: ./node_modules/docx-templates/lib/jsSandbox.js
> Module not found: Error: Can't resolve 'vm' in '/home/weilies_chok/apps/my-app/node_modules/docx-templates/lib'

It seems like someone reported before here but it wasn't supported by the owner of this package (correct me if i am wrong)
https://github.com/guigrpa/docx-templates/issues/86

It's quite a powerful toolkit and i so hunger to integrate the report building into my solution. But these errors making me to re-think is it stable enough for production use. @mathe42  u really helps a lot! even it's still error out

**Comment by alcanrom** [comment]
Hi @weilies, 

I could solve the error adding in tsconfig.json "compilerOptions": {...,"allowSyntheticDefaultImports": true}.

**Comment by weilies** [comment]
> Hi @weilies,
> 
> I could solve the error adding in tsconfig.json "compilerOptions": {...,"allowSyntheticDefaultImports": true}.

It has been enabled while running the script above.

```
  "angularCompilerOptions": {
    "enableI18nLegacyMessageIdFormat": false,
    "allowSyntheticDefaultImports": true,
  }
```

**Comment by ekowalsk** [comment]
> The code is now without syntax error but returned with build error lol
> 
> > Build at: 2021-05-08T10:10:58.912Z - Hash: 57c768fb928842240761 - Time: 2816ms
> > Warning: ./node_modules/sax/lib/sax.js
> > Module not found: Error: Can't resolve 'stream' in '/home/weilies_chok/apps/my-app/node_modules/sax/lib'
> > Error: ./node_modules/docx-templates/lib/jsSandbox.js
> > Module not found: Error: Can't resolve 'vm' in '/home/weilies_chok/apps/my-app/node_modules/docx-templates/lib'
> 
> It seems like someone reported before here but it wasn't supported by the owner of this package (correct me if i am wrong)
> #86
> 
> then i tried myself to npm install vm, it's now changed from build error turn into run-time-error
> 
> > vendor.js:sourcemap:125461 ERROR Error: Uncaught (in promise): Error: Can't find end of central directory : is this a zip file ? If it is, see https://stuk.github.io/jszip/documentation/howto/read_zip.html
> > Error: Can't find end of central directory : is this a zip file ? If it is, see https://stuk.github.io/jszip/documentation/howto/read_zip.html
> 
> here is my latest code
> 
> ```
>   async  postData(url = '', data = {}) {
>     const template: Buffer = await fetch(url).then((v) => v.arrayBuffer() as Promise<any>);
>     const report = await createReport({
>       template,
>       data
>     });
>     this.saveByteArray('myfile.docx', report, true);
>   }
> 
>   onPostData() {
>     let fURL = 'https://firebasestorage.googleapis.com/v0/b/fx-vps-225216.appspot.com/o/tag-example-mod.docx?alt=media&token=3da1cc23-339c-447a-a9e0-e57cc6788155'
>     this.postData(fURL, { answer: 42 })
>   }
> 
>   saveByteArray(reportName: string, byte: ArrayBuffer, word: boolean = false) {
>     const blob = new Blob([byte], {
>       type: (
>         word ? 'vnd.openxmlformats-officedocument.wordprocessingml.document'
>           : 'vnd.openxmlformats-officedocument.spreadsheetml.sheet'
>       )
>     });
>     const link = document.getElementById('doc-download') as HTMLAnchorElement;
>     link.href = window.URL.createObjectURL(blob);
>     link.download = reportName;
>     link.click();
>   }
> ```
> 
> It's quite a powerful toolkit and i so hunger to integrate the report building into my solution. But these errors making me to re-think is it stable enough for production use. @mathe42 u really helps a lot! even it's still error out

I've occurred the  `ERROR Error: Uncaught (in promise): Error: Can't find end of central directory : is this a zip file ?` error when I provided a corrupted docx file

**Comment by weilies** [comment]
The file was corrupted, i pointed to another workable file, also get rid of the CORS issue (they are now accessible by 3rd party site)
here my project https://stackblitz.com/edit/ng-docx-templates-ad47?file=src%2Fapp%2Fapp.component.ts

It's diff error
core.umd.js? [sm]:6539 ERROR Error: Uncaught (in promise): TypeError: Cannot set property 'href' of null
TypeError: Cannot set property 'href' of null
    at AppComponent.saveByteArray (app.component.ts? [sm]:36)
    at AppComponent.eval (app.component.ts? [sm]:20)
    at Generator.next (<anonymous>)
    at fulfilled (tslib.js:114)

**Comment by weilies** [comment]
i put a corrupted doc template. Not sure if it was due to the file manipulated via wps office. But here is the working project
https://stackblitz.com/edit/ng-docx-templates-ad47?file=src%2Fapp%2Fapp.component.ts
it's finally able to generate a word doc, but the PATTEN-ed variable are not replace with value. I am 120% sure the 'data' is provided


```
this.postData(fURL, { name: 'John', surname: 'Appleseed' });
  async postData(url = '', data = {}) {
    const template = await fetch(url).then(
      v => v.arrayBuffer() as Promise<any>
    );

    const report = await createReport({
      template,
      data,
    });
    this.saveByteArray('myfile.docx', report, true);
  }
```

**Comment by jjhbw** [comment]
I assume this is fixed since #216? Feel free to reopen if not.
